events/EventQueue.h

```diff
@@ -938,2465 +725,293 @@ class EventQueue : private mbed::NonCopyable<EventQueue> {
-
-    /** Creates an event bound to the event queue
-     *  @see EventQueue::event
-     */
-    template <typename T, typename R, typename A0, typename A1, typename A2>
-    Event<void(A0, A1, A2)> event(T *obj, R(T::*method)(A0, A1, A2));
-
-    /** Creates an event bound to the event queue
-     *  @see EventQueue::event
-     */
-    template <typename T, typename R, typename A0, typename A1, typename A2>
-    Event<void(A0, A1, A2)> event(const T *obj, R(T::*method)(A0, A1, A2) const);
-
-    /** Creates an event bound to the event queue
-     *  @see EventQueue::event
-     */
-    template <typename T, typename R, typename A0, typename A1, typename A2>
-    Event<void(A0, A1, A2)> event(volatile T *obj, R(T::*method)(A0, A1, A2) volatile);
-
-    /** Creates an event bound to the event queue
-     *  @see EventQueue::event
-     */
-    template <typename T, typename R, typename A0, typename A1, typename A2>
-    Event<void(A0, A1, A2)> event(const volatile T *obj, R(T::*method)(A0, A1, A2) const volatile);
-
-    /** Creates an event bound to the event queue
-     *  @see EventQueue::event
-     */
-    template <typename R, typename A0, typename A1, typename A2>
-    Event<void(A0, A1, A2)> event(mbed::Callback<R(A0, A1, A2)> cb);
-
-    /** Creates an event bound to the event queue
-     *  @see EventQueue::event
-     */
-    template <typename R, typename B0, typename C0, typename A0, typename A1, typename A2>
-    Event<void(A0, A1, A2)> event(R(*func)(B0, A0, A1, A2), C0 c0);
-
-    /** Creates an event bound to the event queue
-     *  @see EventQueue::event
-     */
-    template <typename T, typename R, typename B0, typename C0, typename A0, typename A1, typename A2>
-    Event<void(A0, A1, A2)> event(T *obj, R(T::*method)(B0, A0, A1, A2), C0 c0);
-
-    /** Creates an event bound to the event queue
-     *  @see EventQueue::event
-     */
-    template <typename T, typename R, typename B0, typename C0, typename A0, typename A1, typename A2>
-    Event<void(A0, A1, A2)> event(const T *obj, R(T::*method)(B0, A0, A1, A2) const, C0 c0);
-
-    /** Creates an event bound to the event queue
-     *  @see EventQueue::event
-     */
-    template <typename T, typename R, typename B0, typename C0, typename A0, typename A1, typename A2>
-    Event<void(A0, A1, A2)> event(volatile T *obj, R(T::*method)(B0, A0, A1, A2) volatile, C0 c0);
-
-    /** Creates an event bound to the event queue
-     *  @see EventQueue::event
-     */
-    template <typename T, typename R, typename B0, typename C0, typename A0, typename A1, typename A2>
-    Event<void(A0, A1, A2)> event(const volatile T *obj, R(T::*method)(B0, A0, A1, A2) const volatile, C0 c0);
-
-    /** Creates an event bound to the event queue
-     *  @see EventQueue::event
-     */
-    template <typename R, typename B0, typename C0, typename A0, typename A1, typename A2>
-    Event<void(A0, A1, A2)> event(mbed::Callback<R(B0, A0, A1, A2)> cb, C0 c0);
-
-    /** Creates a
```

Comment:
I believe the variadic form of this would be:

    template <typename F, typename... ContextTypes>
    struct context {
        F f;
        std::tuple<ContextTypes...> c;
        
        context(F f, ContextTypes... c)
            : f(f), c(c...) {}

        template <size_t... I, typename... ArgTs>
        void do_call(std::index_sequence<I...>, ArgTs... args)
        {
            f(get<I>(c)..., args...);
        }
        template <typename... ArgTs>
        void operator()(ArgTs... args)
        {
            do_call(std::index_sequence_for<ContextTypes>(), args...);
        }
    };

But that would involve writing `std::tuple` for ARM C 5.

Reply:
It just requires a small subset of tuple that shouldn't be hard to implement: 

```c++
template<typename... Ts>
struct Tuple : TupleImpl<std:: index_sequence_for <Ts...>, Ts...>;

template<typename Indexes, typename... Ts>
struct TupleImpl; 

template<size_t... Indexes, typename... Ts>
struct TupleImpl<std::index_sequence<Indexes...>, Ts...> : : TupleLeaf<Indexes, Ts>... { };

template<size_t Index, typename T>
struct TupleLeaf { 
    T value;
    T& get();
};

template<size_t Index, typename T, typename... Ts>
struct nth_element_impl {
    using type = typename nth_element_impl<Index-1, Ts...>::type;
};

template<typename T, typename... Ts>
struct nth_element_impl<0, T, Ts...> {
    using type = T;
};

template<size_t Index, typename... Ts>
using nth_element = typename nth_element_impl<Index, Ts...>::type;

template<size_t index, typename Ts...>
auto get(const Tuple<Ts...>& t) -> nth_element<index, Ts...> { 
    return static_cast<TupleLeaf<index, nth_element<index, Ts...>>>(t).get();
}  
```

We can do that latter if it is necessary.

Reply:
Don't see any pressing need here - the non-variadic repetition is small and self-contained, and easier to understand.

But I would appreciate any tips on avoiding the bound argument repetition I've just put back, or the (-,C,V,CV) quads on every member function pointer. Those two multiply together nastily.

Reply:
I think we could use a traits  to test the compatibility between the pointer and the function to bind.
Rules are fairly simple: 
* ptr const: reject non const member function
* ptr volatile: reject non volatile member function

Reply:
Hmm, I guess maybe you get it nearly for free with `is_invocable`. (Having paid the cost there).

As I've implemented that now, I'll look at applying it.

Reply:
It seems to work: https://godbolt.org/z/cS58-7